operator/src/test/java/org/keycloak/operator/testsuite/integration/CacheTest.java

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+
+package org.keycloak.operator.testsuite.integration;
+
+import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.quarkus.logging.Log;
+import io.quarkus.test.junit.QuarkusTest;
+
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusCondition;
+import org.keycloak.operator.crds.v2alpha1.deployment.spec.CacheSpecBuilder;
+import org.keycloak.operator.testsuite.utils.CRAssert;
+
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.keycloak.operator.testsuite.utils.K8sUtils.deployKeycloak;
+
+@QuarkusTest
+public class CacheTest extends BaseOperatorTest {
+
+    private static final String CONFIGMAP_NAME = "my-config";
+
+    @AfterEach
+    public void cleanupConfigMap() {
+        k8sclient.configMaps().withName(CONFIGMAP_NAME).delete();
+    }
+
+    @Test
+    public void testCreateCacheConfigMapFileAfterDeployment() {
+        var kc = getTestKeycloakDeployment(true);
+        var deploymentName = kc.getMetadata().getName();
+        kc.getSpec().setCacheSpec(new CacheSpecBuilder().withNewConfigMapFile("file", CONFIGMAP_NAME, false).build());
+
+        deployKeycloak(k8sclient, kc, false);
+
+        // Check Operator has deployed Keycloak and the statefulset exists, this allows
+        // for the watched configmap to be picked up
+        Log.info("Checking Operator has deployed Keycloak deployment");
+        Resource<StatefulSet> stsResource = k8sclient.resources(StatefulSet.class).withName(deploymentName);
+        Resource<Keycloak> keycloakResource = k8sclient.resources(Keycloak.class).withName(deploymentName);
+        // expect no errors and not ready, which means we'll keep reconciling
+        Awaitility.await().ignoreExceptions().atMost(2, TimeUnit.MINUTES).untilAsserted(() -> {
+            assertThat(stsResource.get()).isNotNull();
+            Keycloak keycloak = keycloakResource.get();
+            CRAssert.assertKeycloakStatusCondition(keycloak, KeycloakStatusCondition.HAS_ERRORS, false);
+            CRAssert.assertKeycloakStatusCondition(keycloak, KeycloakStatusCondition.READY, false);
+        });
+
+        createCacheConfigMap();
+
+        Awaitility.await().ignoreExceptions().atMost(2, TimeUnit.MINUTES).untilAsserted(() -> {
+            assertThat(stsResource.get()).isNotNull();
+            Keycloak keycloak = keycloakResource.get();
+            CRAssert.assertKeycloakStatusCondition(keycloak, KeycloakStatusCondition.HAS_ERRORS, false);
+            CRAssert.assertKeycloakStatusCondition(keycloak, KeycloakStatusCondition.READY, false);
+        });
```

Comment:
I'm not sure if we should be testing like that. No errors and not ready basically happens immediately, doesn't tell us anything about cache config file. If there were errors, they might happen a bit later, not immediately.

Shouldn't we rather wait for Keycloak to be truly ready and also check if the volume mount exists and KC_CACHE_CONFIG_FILE points to it?

Reply:
> I'm not sure if we should be testing like that. No errors and not ready basically happens immediately, doesn't tell us anything about cache config file. If there were errors, they might happen a bit later, not immediately.

It's the same weak check that we're doing in the testKeycloakDeploymentBeforeSecret test. Essentially it's the existence check of the statefulset that we're wanting to see here - that we progressed in resolving enough for that to be created. If there were a RBAC issue, this it won't exist. There is a narrow possibility that the statefulset status may not yet exist, but since we've disabled probes as long as the pods are created successfully that would cause the rollout to be successful. It would be more propper to check that the statefulset is blocked by the configmap, which could be done in the testKeycloakDeploymentBeforeSecret test as well.

> Shouldn't we rather wait for Keycloak to be truly ready and also check if the volume mount exists and KC_CACHE_CONFIG_FILE points to it?

Sorry, yes that was supposed to check for full readiness. I'm not checking the volume mount here because that is covered by a PodTemplate test. We can of course do other things like initially specify an invalid config to make sure that it's being picked up appropriately.

Reply:
Makes sense, thanks.